spec/double_entry/transfer_spec.rb

```diff
@@ -17,13 +17,62 @@ module DoubleEntry
       end
     end
 
+    describe '::transfer' do
+      let(:amount)  { Money.new(10_00) }
+      let(:user)    { User.make! }
+      let(:test)    { DoubleEntry.account(:test, :scope => user) }
+      let(:savings) { DoubleEntry.account(:savings, :scope => user) }
+      let(:new_lines) { Line.all[-2..-1] }
+
+      subject(:transfer) { Transfer.transfer(amount, options) }
+
+      context 'without metadata' do
+        let(:options) { { :from => test, :to => savings, :code => :bonus } }
+
+        it 'creates lines' do
+          expect { transfer }.to change { Line.count }.by 2
+        end
+
+        it 'does not create metadata lines' do
+          expect { transfer }.not_to change { LineMetadata.count }
+        end
+      end
+
+      context 'with metadata' do
+        let(:options) { { :from => test, :to => savings, :code => :bonus, :metadata => { :country => 'AU', :tax => 'GST' } } }
+        let(:new_metadata) { LineMetadata.all[-4..-1] }
+
+        it 'creates metadata lines' do
+          expect { transfer }.to change { LineMetadata.count }.by 4
+        end
+
+        it 'associates the metadata lines with the transfer lines' do
+          transfer
+          expect(new_metadata.count { |meta| meta.line == new_lines.first }).to be 2
+          expect(new_metadata.count { |meta| meta.line == new_lines.last }).to be 2
+        end
+
+        it 'stores the correct metadata' do
+          transfer
+          countries = new_metadata.select { |meta| meta.key == :country }
+          expect(countries.size).to be 2
+          expect(countries.count { |meta| meta.value == 'AU' }).to be 2
+          expect(countries.map(&:line).uniq.size).to be 2
+          taxes = new_metadata.select { |meta| meta.key == :tax }
+          expect(taxes.size).to be 2
+          expect(taxes.count { |meta| meta.value == 'GST' }).to be 2
+          expect(taxes.map(&:line).uniq.size).to be 2
```

Comment:
this spec is rather dense, there are quite a lot of things happening here. Can we tease it apart?